File: epi/Python/list_06.py
# Let h1 and h2 be the heads of lists L1 and L2, respectively. Assume that L1
# & L2 are well-formed, that is each consists of a finite sequence of nodes.
# How would you determine if there exists a node r reachable from both h1 and
# h2 by following the next fields? If such a node exists, find the node that
# appears earliest when traversing the lists. You are constrained to use no
# more than constant additional storage.

def common_node(h1, h2):
    h1_last = find_last(h1)
    h2_last = find_last(h2)
    if h1_last["last"] != h2_last["last"]:
        return None
    distance = abs(h1_last["length"]-h2_last["length"])
    if h1_last["length"] > h2_last["length"]:
        h1 = move_by_k(h1, distance)
    else:
        h2 = move_by_k(h2, distance)
    while True:
        if h1 == h2:
            return h1
        h1 = h1.next
        h2 = h2.next

def find_last(head):
    cur = head
    count = 0
    while cur.next:
        cur = cur.next
        count += 1
    return {"length": count, "last": cur}

def move_by_k(head, k):
    for _ in xrange(k):
        head = head.next
    return head


class Node():
    def __init__(self, data, node):
        self.data = data
        self.next = node

if __name__ == "__main__":
    com = Node(4, Node(5, None))
    h1 = Node(None, Node(1, Node(2, com)))
    h2 = Node(None, Node(3, com))
    print com == common_node(h1, h2)

File: epi/Ruby/list_06.rb
=begin
Let h1 and h2 be the heads of 2 lists respectively. How would you determine if
there exists a node r reachable from both h1 & h2 by following the next fields?
If such a node exists, find the node that appears earliest when traversing the
lists. You are constrained to use no more than constant additional storage.
=end
require "libraries"

def common_node(h1, h2)
  info1 = find_length(h1)
  info2 = find_length(h2)
  if info1[:last] == info2[:last]
    diff = (info1[:count]-info2[:count]).abs
    if info1[:count] > info2[:count]
      h1 = move(h1, diff)
    else
      h2 = move(h2, diff)
    end
    while h1 != h2
      h1 = h1.next
      h2 = h2.next
    end
    h1
  else
    nil
  end
end

def find_length(head)
  count = 0
  last = nil
  while head
    count += 1
    last = head
    head = head.next
  end
  {count: count, last: last}
end

def move(head, k)
  k.times do
    head = head.next if head
  end
  head
end

common = Node.new(5, Node.new(6, Node.new(7, nil)))
h1 = Node.new(1, Node.new(2, Node.new(3, common)))
h2 = Node.new(4, common)
puts common_node(h1, h2).to_s
